python_modules/dagster/dagster/_check/__init__.py

```diff
@@ -1805,3 +1832,201 @@ def _check_two_dim_mapping_entries(
         )  # check level two
 
     return obj
+
+
+# ###################################################################################################
+# ##### CALL BUILDER
+# ###################################################################################################
+
+
+class EvalContext(NamedTuple):
+    global_ns: dict
+    local_ns: dict
+
+    @staticmethod
+    def capture_from_frame(depth: int) -> "EvalContext":
+        ctx_frame = sys._getframe(depth + 1)  # noqa # surprisingly not costly
+
+        return EvalContext(ctx_frame.f_globals, ctx_frame.f_locals)
+
+    def eval_forward_ref(self, ref: ForwardRef) -> Optional[Type]:
+        try:
+            if sys.version_info <= (3, 9):
+                return ref._evaluate(self.global_ns, self.local_ns)  # noqa
+            else:
+                return ref._evaluate(self.global_ns, self.local_ns, frozenset())  # noqa
+        except NameError as e:
+            raise CheckError(f"Unable to resolve {ref}") from e
+
+    def update_from_frame(self, depth: int):
+        ctx_frame = sys._getframe(depth + 1)  # noqa # surprisingly not costly
+        self.global_ns.update(ctx_frame.f_globals)
+        self.local_ns.update(ctx_frame.f_locals)
```

Comment:
Could you add a docblock to describe what is going on here? I think I am inferring quite a bit but docblock would be very helpful.